### PR TITLE
[replica-parallel] Add replica slices concept

### PR DESCRIPTION
**Pull request by gspschmid**
Adds the concept of "replica slices", an explicit representation of which replica ids save which slices out of an array. The intent is for replica slices to allow us to generalize beyond Orbax's current restriction that requires each shard to be saved by exactly one replica.

**Motivation:** Depending on their sharding and replication, JAX arrays may consist of multiple shards. In case of replication each shard carries a distinct `replica_id`, distinguishing the copies of the same logical shard from one another. Orbax's current behavior is to save the same `replica_id`-copy for all shards of all arrays ("single-replica" saving). In the presence of replication this is suboptimal, since the work could be parallelized across all replicas.

This PR is an initial step in the direction of "replica-parallel" saving: we make "replica slices" and related metadata explicit, but do not change any of Orbax's behavior.

Care is taken to compute the resulting `local_shape` (the shape of slices written by each replica) even when the local process does not end up saving any data. This seems necessary, since, to the best of my understanding, only one particular process may set tensorstore metadata, and tensorstore's chunk shape, in particular, is derived from `local_shape`.

### CODE REVIEW

**Comment by gspschmid** [comment]
@cpgaffney1

File: checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+  uniquely responsible for. Slices may be either on-device (as a list of
+  ReplicaSliceOnDevice) or on-host (as a list of indices and numpy arrays).
+  """
+
+  global_shape: Shape
+  local_shape: Shape
+  sharding: jax.sharding.Sharding
+  dtype: np.dtype
+  # Whether the replica slices have been transferred and are ready as ndarrays
+  transferred: bool
```
Comment (cpgaffney1) [inline checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py:62]: Is this really necessary if we can just check `replica_slices` to see whether it contains numpy arrays or not?

File: checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  replica_slices: list[ReplicaSliceOnDevice] | list[tuple[Index, np.ndarray]]
```
Comment (cpgaffney1) [inline checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py:63]: It seems like we could just have a ReplicaSlice object that can store `data: jax.Array | np.ndarray`, with optional replica_id. That would simplify this typing a bit.

**Review by cpgaffney1 — COMMENTED** [review]
Thanks!! Looks good overall, just some minor comments.

File: checkpoint/orbax/checkpoint/_src/serialization/replica_slices_test.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from absl.testing import absltest
+from absl.testing import parameterized
+import jax
+import numpy as np
+from orbax.checkpoint._src.serialization.replica_slices import get_replica_slices, transfer_arrays_to_host
```
Comment (cpgaffney1) [inline checkpoint/orbax/checkpoint/_src/serialization/replica_slices_test.py:19]: Same here, do not directly import functions please.

File: checkpoint/orbax/checkpoint/_src/serialization/replica_slices_test.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+class ReplicaSlicesTest(parameterized.TestCase):
+
+  def setUp(self):
+    super().setUp()
+    num_devices = len(jax.devices())
+    assert num_devices >= 2
+    assert is_pow_of_two(num_devices)
+
+  def test_get_replica_slices_single_replica(self):
```
Comment (cpgaffney1) [inline checkpoint/orbax/checkpoint/_src/serialization/replica_slices_test.py:46]: Would be good to add new cases, or parameterize the existing ones, to include tests for arrays that are not fully replicated.

File: checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py (hunk 1 above)
Comment (gspschmid) [reply checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py]: Because there might be no replica_slices (e.g. because we are in single-replica mode and the current process has no shards with `shard.replica_id == replica_id`)?

File: checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py (hunk 2 above)
Comment (gspschmid) [reply checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py]: Moved from `ReplicaSliceOnDevice` to `ReplicaSlice` (that may be either on-device or on-host) and added some invariants.

File: checkpoint/orbax/checkpoint/_src/serialization/replica_slices_test.py (hunk 4 above)
Comment (gspschmid) [reply checkpoint/orbax/checkpoint/_src/serialization/replica_slices_test.py]: Added a second variant of the test that operates on a partially-replicated array (first dim partitioned).

**Comment by cpgaffney1** [comment]
This is conflicting with a few internal behaviors - I have created a change that fixes most issues, just waiting for some advice on one particular issue. Likely by tomorrow the internal change should be ready to go, at which point I will merge this change and submit mine just after.

**Comment by gspschmid** [comment]
Sounds good, thanks for helping shepherd this through! :-)
